File: 프로그래머스/2/42746. 가장 큰 수/가장 큰 수.js
/* 
주어진 숫자배열에서 순서를 재배치해서 만들수 있는 가장 큰 수 만들기

가장 큰수가 되려면 숫자요소를 문자열로 인식하고 내림차순정렬한 다음 그 요소들을 붙이면 됨
근데 조건 처리를 해줘야함 0~1000이기 때문에 998 999의경우는 ..? -> 조건처리로 푸는 문제가 아닌듯

*/

function solution(numbers) {
  const newNumbers = numbers.map((num) => String(num));
  const resultNum = newNumbers.sort((a, b) => b + a - (a + b));
  return resultNum.every((val) => val == 0) ? "0" : resultNum.join("");
}

console.log(solution([6, 10, 2]));
console.log(solution([3, 30, 34, 5, 9]));
console.log(solution([0, 0, 0, 0, 0]));

File: 프로그래머스/2/42746. 가장 큰 수/가장 큰 수.java
import java.util.*;

class Solution {
    public String solution(int[] numbers) {
       /*
       1. 아이디어
       numbers의 원소들을 문자형으로 변환
       두 원소를 비교할때 가장 왼쪽부터 비교해서 더 큰 값을 내림차순
        3과 30의 경우 3과 3을 비교하고 같으면 3과 0을 비교해서 큰값을 내림차순
       */
        
        String[] strArr= new String[numbers.length];
        
        for(int i=0;i<numbers.length;i++){
            strArr[i] = numbers[i]+"";
        }
        Arrays.sort(strArr, (a,b)->(b+a).compareTo(a+b));
        
        
        StringBuilder sb= new StringBuilder();
        
        for(String s : strArr){
            sb.append(s);
        }
        String result= sb.toString();
        if(result.charAt(0)=='0') return "0";
        else return result;
    }
}